src/clj/witan/send/check_inputs.clj

```diff
@@ -28,15 +29,54 @@
       (set all-states)
       (set (map #(dissoc % :academic-year) all-states)))))
 
+(defn joiner? [age transition]
+  (and (= (:academic-year-2 transition) age) (= (:setting-1 transition) :NONSEND)))
+
+(defn leaver? [age transition]
+  (and (= (:academic-year-1 transition) age) (= (:setting-2 transition) :NONSEND)))
+
+(defn check-all-ages-present [transitions]
+  "Checks all ages are represented in both academic years 1 and 2"
+  (let [transitions-to-ages (distinct (map :academic-year-2 transitions))
+        transitions-from-ages (distinct (map :academic-year-1 transitions))
+        all-ages (sort (distinct (concat transitions-to-ages transitions-from-ages)))
+        possible-ages (range (first all-ages) (+ 1 (last all-ages)))]
+    (doseq [a (drop 1 possible-ages)]
+      (when-not (some #(= a %) transitions-to-ages)
+        (do (r/info (r/bold (str "There are no transitions to AY " a)))
+            (repl-warn (str "There are no transitions to AY " a)))))
+    (doseq [a (drop-last 1 possible-ages)]
+      (when-not (some #(= a %) transitions-from-ages)
+        (do (r/info (r/bold (str "There are no transitions from AY " a)))
+            (repl-warn (str "There are no transitions from AY " a)))))))
+
+(defn check-joiner-leaver-gaps [transitions]
+  "bring in range of ages in dataset, check that all ages have joiners and leavers"
+  (let [ages (sort (distinct (concat (map :academic-year-2 transitions) (map :academic-year-1 transitions))))]
+    (doseq [a (drop-last 1 ages)]
+      (when (empty? (filter #(joiner? a %) transitions))
+        (do (r/info (r/bold (str "There are no joiners to AY " a)))
+            (repl-warn (str "There are no joiners to AY " a)))))
+    (doseq [a (drop 1 ages)]
+      (when (empty? (filter #(leaver? a %) transitions))
+        (do (r/info (r/bold (str "There are no leavers from AY " a)))
+            (repl-warn (str "There are no leavers from AY " a)))))))
```

Comment:
same as above around separating the checks from the reporting so we can test the checks